### ISSUE TITLE
Number of files covered before and after using deep-cover is different

### ISSUE THREAD

**Issue by zs787** [deep-cover/deep-cover#50]
Hi,

We have recently added deep-cover and we are already using simplecov so we did the settings according to the take over section in the read-me. 

When looking at the results, in the generated coverage report I see a difference in the number of files covered. Following is an example:

**With deep-cover**
```
All Files (95.97% covered at 1195.65 hits/line)
105 files in total. 3478 relevant lines. 3338 lines covered and 140 lines missed
```
**Without deep-cover**
```
All Files (99.98% covered at 342.39 hits/line)
219 files in total. 13312 relevant lines. 13310 lines covered and 2 lines missed
```
Another scenario on another application we get only 3 non src files covered, when using deep-cover take over.

**Versions used**
```
Ruby version 2.5.3
deep-cover 0.7.3 
simplecov 0.16.1
```

deep-cover usage
```
if ENV['COVERAGE']
  require 'deep_cover/builtin_takeover'
  require 'simplecov'
  require 'codecov'

  SimpleCov.formatter = SimpleCov::Formatter::MultiFormatter.new(
    [
      SimpleCov::Formatter::Codecov,
      SimpleCov::Formatter::HTMLFormatter
    ]
  )

  SimpleCov.start do
    ...
  end
end
```

Any idea if this is expected behaviour or do we have to change anything in the configs to make this work as expected ?  As we expect the number of files covered to be the same, but the coverage rate to be different. 

Thanks

**Comment by marcandre** [comment]
If I'm not mistaken, `simple_cov` will only cover your app by default, so we should have the same defaults, so yes, the same files should be covered.

Just to be sure, the `require_relative 'config/application'` is called (for the first time) after the `require 'deep_cover/builtin_takeover'`, right? Note that the standard rails `Rakefile` loads the application, so if you're using `rake` that could be an issue.

**Comment by marcandre** [comment]
BTW, I'm traveling this week, so I don't have the time to make a dummy Rails app and verify it works as expected though, maybe @MaxLap can check it out?

**Comment by MaxLap** [comment]
A missing thing about takeover is that right now, it will only cover files which match deep-cover's configuration.

By default, for a Rails app, this means the content of the `app/` and of the `lib/` directories.

If this is not a Rails app, then it will cover everything in the directory except these directories: `autotest/ features/ spec/ test/ bin/ exe/`.

Can you try adding the following after the `require 'deep_cover/builtin_takeover'` helps?
```
DeepCover.configure do
  paths Dir['./*/'] - %w(./autotest/ ./features/ ./spec/ ./test/ ./bin/ ./exe/)
end
```

The configuration you have in `SimpleCov.start` could be helpful to see if some other settings are causing a difference.

**Comment by zs787** [comment]
Thanks guys for your responses.

- Our app is a rails app and when using deep-cover on specific app we get only 3 files under `config/` covered where it should be 44 files, including what is under `app/` at least, which is all covered by default by just using simplecov.
- Taking out all configs under the `SimpleCov.start` still wont make any difference, for the scenario above. The configs here are just grouping couple of paths.
- Thats true it will cover what is under `app/` and `lib/` but for example in one app without deep-cover and with only simplecov we have files under `config/` covered, but once we have deep-cover added then we lose coverage for those files.
- All results are the same where im looking at the locally generated file or in our pipeline I get the same thing, testing with and without deep-cover.
- In summary what is being covered by simplecov is not all covered by deep-cover. Any suggestions would be appreciated.

**Comment by zs787** [comment]
Hi,

Just wondering if any of you guys were able or had the time to reproduce this issue ?
We have previously investigated the results of deep-cover and we like the coverage criteria, but we need to be able to have it covering everything we cover with ```simplecov``` before we proceed.

Thank you

**Comment by MaxLap** [comment]
Not yet, sorry. Bad timing, we have both been travelling.

Its not clear from your last post, did you try my suggestion to add this configuration? 

```
DeepCover.configure do
  paths Dir['./*/'] - %w(./autotest/ ./features/ ./spec/ ./test/ ./bin/ ./exe/)
end
```

**Comment by zs787** [comment]
> Its not clear from your last post, did you try my suggestion to add this configuration?

no worries and yes I did try it and nothing changed, I was still getting less than the acutal number of files we get with simplecov.

**Comment by marcandre** [comment]
It's though to imagine how/why that could happen.

@MaxLap one thing we should do in `takeover` is double check that `BuiltinCoverage.running?` is `false`, just to make sure it's not something like `SimpleCov.start; ...; require 'deep_cover/builtin_takeover'; SimpleCov.start`, right?

**Comment by zs787** [comment]
Hi guys, I understand you both been busy so just wondering if any of you were able to reproduce this issue at all ?

Thanks

**Comment by MaxLap** [comment]
Hello, I haven't been able to reproduce your issue. I use takeover just fine in some of my projects.

I committed a possible fix for the issue, however, we are having an issue with Travis I have no idea why, but it is currently failing, which is making me uneasy about doing a release. We haven't had to time to investigate that Travis issue yet.

Because DeepCover is actually 2 gems in the same Github repository, I have no idea if you can point your Gemfile the master branch on Github.

As for reproducing, if you could basically duplicate your project and just remove everything in order to keep a single file that should be covered but isn't, it would be really helpful and shouldn't be too complicated. Just replace its content with a very simple thing, make sure it gets required and wipe everything else.

**Comment by MaxLap** [comment]
Hello, finally got around to fixing travis. I just released 0.7.4 which I hope solves your issue.

If it doesn't, then we will really need your help with providing us a reproduction case.

**Comment by zs787** [comment]
@MaxLap thanks for the fix. So far seems to be working I'm continuing to run it on all previously known apps to have this issue to conclude my testing. 

However, one new error I'm seeing and only for two files is this:
`Can't cover /path/to/file/xxxi_spec.rb because of incompatible encoding (see issue #9)`
haven't had a chance to dig deeper but since im writing here I thought I would ask because I only see it when enabling deep-cover

**Comment by MaxLap** [comment]
Glad things seem to be working out.

That message is really not clear.. Basically, the parser we use for deep-cover can't parse that file, so we can't instrument it to do coverage. See: https://github.com/whitequark/parser#invalid-characters-inside-comments-and-literals

The change I made now makes DeepCover even process test files when running in takeover mode, this is why you didn't see that warning in the past.

**Comment by marcandre** [comment]
Hopefully the file will be easily fixed, or else it doesn't matter too much that it isn't covered.

**Comment by zs787** [comment]
everything looks ok and thanks for the fix.

**Comment by MaxLap** [comment]
Great news!
